src/BizHawk.Emulation.Cores/Consoles/Nintendo/BSNES/BsnesCore.cs

```diff
@@ -0,0 +1,391 @@
+			{
+				case "cx4": firmwareId = "CX4"; break;
+				case "dsp1": firmwareId = "DSP1"; break;
+				case "dsp1b": firmwareId = "DSP1b"; break;
+				case "dsp2": firmwareId = "DSP2"; break;
+				case "dsp3": firmwareId = "DSP3"; break;
+				case "dsp4": firmwareId = "DSP4"; break;
+				case "st010": firmwareId = "ST010"; break;
+				case "st011": firmwareId = "ST011"; break;
+				case "st018": firmwareId = "ST018"; break;
+				default:
+					CoreComm.ShowMessage($"Unrecognized SNES firmware request \"{hint}\".");
+					return "";
+			}
+
+			string ret = "";
+			var data = CoreComm.CoreFileProvider.GetFirmware("SNES", firmwareId, required, "Game may function incorrectly without the requested firmware.");
+			if (data != null)
+			{
+				ret = hint;
+				Api.AddReadonlyFile(data, hint);
+			}
+
+			Console.WriteLine("Served bsnescore request for firmware \"{0}\"", hint);
+
+			// return the path we built
+			return ret;
+		}
+
+		private enum LoadParamType
+		{
+			Normal, SuperGameBoy
+		}
+
+		private struct LoadParams
+		{
+			public LoadParamType type;
+			public string baseRomPath;
+			public byte[] romData;
+			public byte[] sgbRomData;
+		}
+
+		private void LoadCurrent()
+		{
+			if (_currLoadParams.type == LoadParamType.Normal)
+				Api.core.snes_load_cartridge_normal(_currLoadParams.baseRomPath, _currLoadParams.romData, _currLoadParams.romData.Length);
+			else
+				Api.core.snes_load_cartridge_super_gameboy(_currLoadParams.baseRomPath, _currLoadParams.romData,
+					_currLoadParams.sgbRomData, (ulong) _currLoadParams.romData.Length << 32 |  (uint)_currLoadParams.sgbRomData.Length);
+
+			_region = Api.core.snes_get_region();
+		}
+
+		// poll which updates the controller state
+		private void snes_input_poll()
+		{
+			_controllers.CoreInputPoll(_controller);
+		}
+
+		/// <param name="port">0 or 1, corresponding to L and R physical ports on the snes</param>
+		/// <param name="index">meaningless for most controllers.  for multitap, 0-3 for which multitap controller</param>
+		/// <param name="id">button ID enum; in the case of a regular controller, this corresponds to shift register position</param>
+		/// <returns>for regular controllers, one bit D0 of button status.  for other controls, varying ranges depending on id</returns>
+		private short snes_input_state(int port, int index, int id)
+		{
+			return _controllers.CoreInputState(port, index, id);
+		}
+
+		private void snes_no_lag()
+		{
+			// gets called whenever there was input polled, aka no lag
+			IsLagFrame = false;
+		}
+
+		private readonly int[] palette = new int[32768];
+
+		private void generate_palette()
+		{
+			for (int color = 0; color < 32768; color++) {
+				int r = (color >> 10) & 31;
+				int g = (color >>  5) & 31;
+				int b = (color >>  0) & 31;
+
+				r = r << 3 | r >> 2; r = r << 8 | r << 0;
+				g = g << 3 | g >> 2; g = g << 8 | g << 0;
+				b = b << 3 | b >> 2; b = b << 8 | b << 0;
+
+				palette[color] = r >> 8 << 16 | g >> 8 <<  8 | b >> 8 << 0;
```

Comment:
Back in the day, some of our frontend code required the top 8 bits of this to be 1.  That may or may not be the case now?  I forget.